datanode/datanode.go

```diff
@@ -302,83 +306,140 @@ func (d *dataNode) startActiveTopologyWatch() {
 	}
 }
 
-func (d *dataNode) startAnalyzingShardAvailability() {
-	ticker := time.NewTicker(5 * time.Second)
+func (d *dataNode) checkShardReadiness(tables []string, shards []uint32) (numReadyShards int) {
+	// check all whether all tables within initialing shards are bootstrapped
+	for i := 0; i < len(shards); i++ {
+		shardID := shards[i]
+		numTablesBootstrapped := 0
+		for _, table := range tables {
+			tableShard, err := d.memStore.GetTableShard(table, int(shardID))
+			if err != nil {
+				d.logger.With(
+					"error", err.Error(),
+					"table", table,
+					"shard", shardID).
+					Error("cannot get table shard")
+				continue
+			}
+			if tableShard.IsBootstrapped() {
+				numTablesBootstrapped++
+			}
+			tableShard.Users.Done()
+		}
+
+		if numTablesBootstrapped == len(tables) {
+			shards[i], shards[numReadyShards] = shards[numReadyShards], shards[i]
```

Comment:
what purpose of this line?

Reply:
swap two element

this is part of the partition algorithm, which partitions the list into two parts, ready shards and non ready shards